src/mysql.rs

```diff
@@ -63,4 +136,303 @@ impl MySQLTableFactory {
+            .map_err(to_datafusion_error)?;
+
+        let primary_keys = get_primary_keys_from_constraints(&cmd.constraints, &schema);
+
+        mysql
+            .create_table(Arc::clone(&schema), &mut transaction, primary_keys)
+            .await
+            .map_err(to_datafusion_error)?;
+
+        for index in indexes {
+            mysql
+                .create_index(
+                    &mut transaction,
+                    index.0.iter().collect(),
+                    index.1 == IndexType::Unique,
+                )
+                .await
+                .map_err(to_datafusion_error)?;
+        }
+
+        transaction
+            .commit()
+            .await
+            .context(UnableToCommitMySQLTransactionSnafu)
+            .map_err(to_datafusion_error)?;
+
+        drop(conn_guard);
+
+        let dyn_pool: Arc<DynMySQLConnectionPool> = pool;
+
+        let read_provider = Arc::new(SqlTable::new_with_schema(
+            "mysql",
+            &dyn_pool,
+            Arc::clone(&schema),
+            TableReference::bare(name.clone()),
+            Some(Engine::MySQL),
+        ));
+
+        let read_provider = Arc::new(read_provider.create_federated_table_provider()?);
+        Ok(MySQLTableWriter::create(read_provider, mysql, on_conflict))
+    }
+}
+
+pub struct MySQL {
+    table_name: String,
+    pool: Arc<MySQLConnectionPool>,
+    schema: SchemaRef,
+    constraints: Constraints,
+}
+
+impl MySQL {
+    #[must_use]
+    pub fn new(
+        table_name: String,
+        pool: Arc<MySQLConnectionPool>,
+        schema: SchemaRef,
+        constraints: Constraints,
+    ) -> Self {
+        Self {
+            table_name,
+            pool,
+            schema,
+            constraints,
+        }
+    }
+
+    #[must_use]
+    pub fn table_name(&self) -> &str {
+        &self.table_name
+    }
+
+    #[must_use]
+    pub fn constraints(&self) -> &Constraints {
+        &self.constraints
+    }
+
+    pub async fn connect(&self) -> Result<Box<DynMySQLConnection>> {
+        let mut conn = self.pool.connect().await.context(DbConnectionSnafu)?;
+
+        let mysql_conn = Self::mysql_conn(&mut conn)?;
+
+        if !self.table_exists(mysql_conn).await {
+            TableDoesntExistSnafu {
+                table_name: self.table_name.clone(),
+            }
+            .fail()?;
+        }
+
+        Ok(conn)
+    }
+
+    pub fn mysql_conn(db_connection: &mut Box<DynMySQLConnection>) -> Result<&mut MySQLConnection> {
+        let conn = db_connection
+            .as_any_mut()
+            .downcast_mut::<MySQLConnection>()
+            .context(UnableToDowncastDbConnectionSnafu)?;
+
+        Ok(conn)
+    }
+
+    async fn table_exists(&self, mysql_connection: &MySQLConnection) -> bool {
+        let sql = format!(
+            r#"SELECT EXISTS (
+          SELECT 1
+          FROM information_schema.tables
+          WHERE table_name = '{name}'
+        )"#,
+            name = self.table_name
```

Comment:
Is it possible to use the parameterization offered by the mysql crate to pass in the table name as a parameter? Otherwise this could be susceptible to a SQL injection attack. (I know the other providers follow this pattern, but its on my todo list to change)

Reply:
Or even using sea_query to build it like the other methods do

Reply:
Hi @phillipleblanc,

Thank you for your suggestion regarding SQL security and using `sea-query`! During implementation, I identified some limitations with handling table names in multi-level database structures, so I explored an alternative approach on a separate branch. Here’s the full background and the design changes I considered:

### Background and Motivation

Initially, the goal was to replace raw SQL string formatting with `sea-query` for enhanced security. However, I discovered that the string-based table names had certain limitations, particularly in multi-level database configurations:

1. **Limitations of `information_schema.tables` Queries**: When users connect to a default database (e.g., `mysql://user1:password@localhost/database1`), they can access tables in that database with a simple table name (e.g., `table_name`). However, accessing tables in other databases (e.g., `database2.table_name`) requires a fully qualified name, which `information_schema.tables` cannot handle. This leads to ambiguity with simple table names and, for default database tables, requires extracting the database name from the connection configuration, adding complexity.

2. **Simplifying Logic with `SELECT * FROM table_name LIMIT 0`**: To avoid the limitations of `information_schema.tables`, I explored an alternative approach on a separate branch that uses `SELECT * FROM table_name LIMIT 0` to verify table existence. This method leverages the default database setting of the connection, allowing any valid table name format to be used. The user’s configuration dictates the table name format, removing the need for extra parsing.

3. **Seamless Integration with `sea-query`**: In this branch, I replaced the `String` table name type with `TableReference` and used a `table_reference_to_table_ref` function to convert it to `TableRef`. This approach naturally supports multi-level names while maintaining clarity and security.

### Suggested Next Steps

To keep the current PR focused and limit changes, my suggestion is not to modify the SQL construction logic here but to implement the write functionality as planned. Once this PR is merged, I’ll perform a full refactor on a separate branch using `sea-query` to support multi-level table names as described above.You can check out the changes I’ve made so far on this branch: [table_ref branch](https://github.com/gengteng/datafusion-table-providers/tree/table_ref).

Looking forward to your feedback, and thank you again for helping to refine these design ideas! 😊

Reply:
Apologies for the late reply! Thank you for your detailed explanation, makes sense to me. Once the tests are all passing, I think this PR will be good to go. Thanks for your patience!

Reply:
@gengteng https://github.com/datafusion-contrib/datafusion-table-providers/pull/176 was merged, which bumps this project to DF 43 - so you should be able to bring in your federation change now.